### PR TITLE
Parse names out of resource directory entries.

### PR DESCRIPTION
**Pull request by wxsBSD**
If the Name field has the high bit set the remaining 31 bits are an
offset relative to the start of the resource section. At that offset is
a resource directory string, which contains 2 bytes for the length and a
variable length unicode string.

Teach the resource parser about these by calling a function which will
return a pointer to the resource directory string or NULL if there is no
string or the entire string is not available.

The callback - pe_collect_resources() - takes the pointers to the
strings and calls set_sized_string() on them so they are exposed to
rules.

While I'm here consistently use semicolons in the declare_* lines.

### CODE REVIEW

**Comment by wxsBSD** [comment]
Here's a rule that shows how to use this:

```
import "pe"

// Look for a resource named "SCREEN0" in Unicode.
rule name_test {
  condition:
    for any i in (0..pe.number_of_resources - 1):
      (pe.resources[i].name_string == "S\x00C\x00R\x00E\x00E\x00N\x000\x00")
} 

// Look for FILE in Unicode.
rule type_test {
  condition:
    for any i in (0..pe.number_of_resources - 1):
      (pe.resources[i].type_string == "F\x00I\x00L\x00E\x00")
}
```

I ran this against 587 random samples I had laying around and it found 19.

Here's the output (all the samples are on VT):

```
type_test /Users/wxs/Desktop/malware/signed_bins/random/01ee9fa7202c5afa8beff766e4f9d694
name_test /Users/wxs/Desktop/malware/signed_bins/random/0cfd6bae475be1023f94bd5c0a5b4e6a
name_test /Users/wxs/Desktop/malware/signed_bins/random/0b1ec85ebb4d23250c70524e135c5732
name_test /Users/wxs/Desktop/malware/signed_bins/random/303390172ac7ce61dada40a3db768e8f
name_test /Users/wxs/Desktop/malware/signed_bins/random/446da9bf8a96755887105668e80babca
name_test /Users/wxs/Desktop/malware/signed_bins/random/5c4d88c4bdb2b61b0ce465f3436dada8
type_test /Users/wxs/Desktop/malware/signed_bins/random/688cafaff9f228c41b27422abd7d0d4e
type_test /Users/wxs/Desktop/malware/signed_bins/random/715639806101323dec6cfe43a8dd13df
name_test /Users/wxs/Desktop/malware/signed_bins/random/827ebeac41c19c67ca90651104f73d93
type_test /Users/wxs/Desktop/malware/signed_bins/random/93c7f7b610cdced1526370ae3a56cd11
name_test /Users/wxs/Desktop/malware/signed_bins/random/b8d91db370d21c3bbd4142b0e2c64456
name_test /Users/wxs/Desktop/malware/signed_bins/random/c8c6e27b1b65c7b65c6c2a0ea98d0afd
type_test /Users/wxs/Desktop/malware/signed_bins/random/ca9cac94fe86e93893a0634b18074b8c
name_test /Users/wxs/Desktop/malware/signed_bins/random/cd7ba9f900389bbc12f2067244576da9
name_test /Users/wxs/Desktop/malware/signed_bins/random/d1222d6301ed72aa277c5964ce6a8a79
name_test /Users/wxs/Desktop/malware/signed_bins/random/d61d178ef09bd38e3cab6aefff65a625
name_test /Users/wxs/Desktop/malware/signed_bins/random/e4f800d88764b928131abc3ed5e5e79a
name_test /Users/wxs/Desktop/malware/signed_bins/random/e4da47b5d6d24a436531003d1c9967f7
type_test /Users/wxs/Desktop/malware/signed_bins/random/e6ff6a9a384b218104b9513a26295bdb
````

File: libyara/modules/pe.c [hunk 1]
```diff
@@ -1532,11 +1613,14 @@ begin_declarations;
   declare_integer("resource_major_version")
```
Comment (plusvic) [inline libyara/modules/pe.c]: To be coherent with linker_version, os_version, image_version and subsystem_version, declare an struct "resource_version" containing two fields "major" and "minor".

**Comment by wxsBSD** [comment]
Also, if you don't mind I'm willing to quickly add all the resource type constants to this module.

It would make checking for specific resource types much easier than having to know the values.

**Comment by wxsBSD** [comment]
What is your opinion on not setting ```type```, ```name```, or ```id``` if the corresponding string is not NULL? If the string is not NULL then the integers are meaningless and should probably stay UNDEFINED.

**Comment by plusvic** [comment]
Yes, I agree with that, if they are strings then the integers don't make
sense.

**Comment by plusvic** [comment]
Manually merged in b0d0494dc635e34ea7ad06530f0d156801cf3b50

**Pull request closed without merging**